go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -0,0 +1,239 @@
+	// the second the GW IP (next hop), and the third the GR name
+	PodExternalRoutes map[string]map[string]string
+}
+
+type routePolicy struct {
+	labelSelector   *metav1.LabelSelector
+	staticGateways  gatewayInfoList
+	dynamicGateways map[ktypes.NamespacedName]*gatewayInfo
+}
+
+type externalPolicyManager struct {
+	stopCh <-chan struct{}
+
+	// route policies
+
+	routeLister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister
+
+	// Pods
+	podLister corev1listers.PodLister
+
+	// Namespaces
+	namespaceLister corev1listers.NamespaceLister
+
+	// cache for set of policies impacting a given namespace
+	namespaceInfoCache      map[string]*namespaceInfo
+	namespaceInfoCacheMutex *sync.Mutex
+	namespaceLockMap        map[string]*sync.Mutex
+	namespaceLockMapMutex   *sync.Mutex
+
+	// cache that captures the current route policies, with their processed information
+	routePolicyCache      map[string]*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute
+	routePolicyCacheMutex *sync.Mutex
+	netClient             networkClient
+}
+
+func newExternalPolicyManager(
+	stopCh <-chan struct{},
+	podLister corev1listers.PodLister,
+	namespaceLister corev1listers.NamespaceLister,
+	routeLister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister,
+	netClient networkClient) *externalPolicyManager {
+
+	m := externalPolicyManager{
+		stopCh:                  stopCh,
+		routeLister:             routeLister,
+		podLister:               podLister,
+		namespaceLister:         namespaceLister,
+		namespaceInfoCache:      make(map[string]*namespaceInfo),
+		namespaceInfoCacheMutex: &sync.Mutex{},
+		namespaceLockMap:        make(map[string]*sync.Mutex),
+		namespaceLockMapMutex:   &sync.Mutex{},
+		routePolicyCache:        make(map[string]*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute),
+		routePolicyCacheMutex:   &sync.Mutex{},
+		netClient:               netClient,
+	}
+
+	return &m
+}
+
+func (m *externalPolicyManager) getRoutePolicyFromCache(policyName string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, bool) {
+	m.routePolicyCacheMutex.Lock()
+	defer m.routePolicyCacheMutex.Unlock()
+	p, found := m.routePolicyCache[policyName]
+	return p, found
+}
+
+func (m *externalPolicyManager) updateRoutePolicyInCache(policy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) {
+	m.routePolicyCacheMutex.Lock()
+	defer m.routePolicyCacheMutex.Unlock()
+	m.routePolicyCache[policy.Name] = policy
+}
+
+func (m *externalPolicyManager) deleteRoutePolicyFromCache(policyName string) {
+	m.routePolicyCacheMutex.Lock()
+	defer m.routePolicyCacheMutex.Unlock()
+	delete(m.routePolicyCache, policyName)
+}
+
+// GetDynamicGatewayIPsForTargetNamespace is called by the annotation logic to identify if a namespace is managed by an CR.
+// Since the call can occur outside the lifecycle of the controller, it cannot rely on the namespace info cache object to have been populated.
```

Comment:
well couldn't you first check if the cache is populated or not? If it is then you can shortcut?

Reply:
I don't know when the cache has been completely populated or is in the process of. I could add another field that would give me that information. What do you think?